CMakeLists.txt

```
@@ -1,17 +1,14 @@
 # Icinga 2 | (c) 2012 Icinga GmbH | GPLv2+
 
-cmake_minimum_required(VERSION 2.8.12)
+# The CMake policies starting from 3.0 to 3.17 shouldn't affect the Icinga 2 build process in any way,
+# so the maximum CMake required version is set to 3.17. Anything above that may work, but we didn't check
+# the policies introduced in CMake 3.18 and later.
+cmake_minimum_required(VERSION 3.8...3.17)
```

> > @julianbrost The burden is that comment one has to write/read/understand **in order to** reflect "the state of the CMake config with this PR more accurately".
> 
> That sounds like you just hate the `...` syntax and you wouldn't even be happy with `cmake_minimum_required(VERSION 3.16...3.17)`.
> 
> I think it makes sense to say "we verified the policies up to 3.17 and they work fine" so that if 3.17 is used, it doesn't have to emulate

On the other side, after I thought about it, I'd indeed _prefer_ one version, i.e one policy set, not even for the syntax: IMAO we already have enough stuff which is/behaves differently across OS. Compilers, Boost, ... Also it's pretty cool to compile stuff on the own workstation with the latest CMake and being sure GHA will work as well, due to that one CMake version forcing my toolchain to behave exactly like everything we support:

```suggestion
cmake_minimum_required(VERSION 3.16.3)
```

And that "emulation" shouldn't be a big deal.

---

> Also it's pretty cool to compile stuff on the own workstation with the latest CMake and being sure GHA will work as well, due to that one CMake version forcing my toolchain to behave exactly like everything we support:

Does it behave exactly the same? We are currently doing this PR because a new CMake version doesn't do that. And it's a compatibility mode, not a "download and run the old version", so there can be differences nonetheless, including bugs that are present only in some versions (like for example what Yonas just happend to discover in https://github.com/Icinga/icinga2/pull/10403#issuecomment-2777830186). The CI exists to test for compatibility, not your local machine.

---

> We are currently doing this PR because a new CMake version doesn't do that. And it's a compatibility mode,

Actually, because it dropped the compatibility mode.

> including bugs that are present only in some versions (like for example what Yonas just happend to discover in https://github.com/Icinga/icinga2/pull/10403#issuecomment-2777830186

Yes, very unfortunately even compatibility isn't perfect.

> The CI exists to test for compatibility, not your local machine.

And to prevent wasting time on my local machine and discovering latter by CI, I build and test with the oldest possible stuff (which already paid off btw:

* https://github.com/Icinga/icinga-notifications/pull/226#discussion_r1674287227)

@yhabteab Given all above arguments, please choose one:

```suggestion
cmake_minimum_required(VERSION 3.16.3)
```

```suggestion
cmake_minimum_required(VERSION 3.16.3...3.17)
```

---

> And to prevent wasting time on my local machine and discovering latter by CI, I build and test with the oldest possible stuff (which already paid off btw:

That's nice for you and nobody is stopping you from building with CMake 3.16 locally if that's what you'd like to do.

["Because I can"](https://github.com/Icinga/icinga2/pull/10402#discussion_r2027251278) is just a weak argument for raising that requirement.

I mean I also don't propose adding "reject building if OS is neither Linux nor Windows" to our CMake config "just because I can" and we [don't officially build for/support other operating systems](https://icinga.com/subscriptions/support-matrix/) like your beloved Mac OS and OpenBSD.

---

> ["Because I can"](https://github.com/Icinga/icinga2/pull/10402#discussion_r2027251278) is just a weak argument for raising that requirement.

Ok, here is a better argument:

* "CMake supports unity builds starting with version 3.16" (#10404)

> I mean I also don't propose adding "reject building if OS is neither Linux nor Windows" to our CMake config "just because I can" and we [don't officially build for/support other operating systems](https://icinga.com/subscriptions/support-matrix/) like your beloved Mac OS and OpenBSD.

Of course we don't require a higher OS version _for nothing,_ hence my argument **actually** was:

> I require higher versions of stuff just because I can. **Except where I can't [because e.g a supported OS], that's the limit then** – in this case v3.16.3.

But if all OS versions we support already have tool X version Y+, I don't see why we should support lover versions of tool X. **There** I consider "just because I can" a completely sufficient argument.

---

> > ["Because I can"](https://github.com/Icinga/icinga2/pull/10402#discussion_r2027251278) is just a weak argument for raising that requirement.
> 
> Ok, here is a better argument:
> 
> * "CMake supports unity builds starting with version 3.16" ([Use CMake unity builds instead of custom implementation #10404](https://github.com/Icinga/icinga2/issues/10404))

Then this PR raises the minimum because it has a reason for doing so, no problem.

> > I mean I also don't propose adding "reject building if OS is neither Linux nor Windows" to our CMake config "just because I can" and we [don't officially build for/support other operating systems](https://icinga.com/subscriptions/support-matrix/) like your beloved Mac OS and OpenBSD.
> 
> Of course we don't require a higher OS version _for nothing,_ hence my argument **actually** was:
> 
> > I require higher versions of stuff just because I can. **Except where I can't [because e.g a supported OS], that's the limit then** – in this case v3.16.3.

That's exactly the same argument (otherwise you'd have to ask for `cmake_minimum_required(VERSION 4.0)`).

> But if all OS versions we support already have tool X version Y+, I don't see why we should support lover versions of tool X. **There** I consider "just because I can" a completely sufficient argument.

I don't.

---

So, just to be clear, you're keeping v3.8 support for now because, in your opinion, there aren't sufficient reasons against? And support means, like, it should work and we're responsible for latter?

## "Fun" fact

* If e.g we #10404 on top of this PR, but don't raise cmake_minimum_required, GHA wouldn't complain, according to https://stackoverflow.com/questions/73712129/how-to-detect-using-too-new-features-in-cmake?utm_source=chatgpt.com

This means, we'd "support" CMake v3.8 on the paper, but nothing (automatically) tests that.

---

Support in the sense that we don't intentionally break it for no good reason. Just like we don't add "if OpenBSD then fatal error" for no particular reason even though we don't fully support it nor have automatic tests for that either.